### ISSUE TITLE
False positives with `impl Iterator<>` and `iter_not_returning_iterator`

### ISSUE THREAD

**Issue by repi** [rust-lang/rust-clippy#8293]
### Summary

Tested the `iter_not_returning_iterator` with latest nightly on our codebase and ran into a couple of cases that I think are false positives. Both of these functions are called `iter` but do implement iterator, though `impl Iterator<>`

### Lint Name

iter_not_returning_iterator

### Reproducer

I tried this code:

```rust
struct Test1 {
    values: std::collections::VecDeque<(f64, f32)>,
}

impl Test1 {
    pub fn iter(&'_ self) -> impl Iterator<Item = (f64, f32)> + '_ {
        self.values.iter().map(|(a, b)| (*a, *b))
    }
}

struct Test2 {
    values: std::collections::BTreeSet<String>,
}

impl Test2 {
    pub fn iter(&self) -> impl Iterator<Item = &String> {
        self.values.iter()
    }
}
```

I saw this happen:

```
warning: this method is named `iter` but its return type does not implement `Iterator`
 --> clippy-repro/src/main.rs:6:5
  |
6 |     pub fn iter(&'_ self) -> impl Iterator<Item = (f64, f32)> + '_ {
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: requested on the command line with `-W clippy::iter-not-returning-iterator`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_not_returning_iterator

warning: this method is named `iter` but its return type does not implement `Iterator`
  --> clippy-repro/src/main.rs:16:5
   |
16 |     pub fn iter(&self) -> impl Iterator<Item = &String> {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_not_returning_iterator
```

I expected to see this happen:

Both of these functions do implement `Iterator` so lint shouldn't trigger on them

### Version

```text
clippy 0.1.60 (ad46af24 2022-01-14)
```

**Comment by Jarcho** [comment]
Fixes in #8287

**Comment by repi** [comment]
Awesome thanks!
